rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+
+- ⚠ We can mistakenly debug an issue against the undeployed version of the code.
+- ⚠ We may struggle to isolate a breaking change, if many were deployed together.
+
+## Proposal
+
+Make deployments fully automatic for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145). This has already begun, as part of [a successful trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
```

Comment:
I have a few questions which I'm unsure of where they might fit into the rest of this RFC, so I'll put them here under safety concerns:

- I feel that there is some general consensus that releasing changes late in the week (think an hour or so before the end of the working week) is something that we generally put off until we're back for the next week, unless the change is urgent. Given that it's going to be much easier to press the merge button on a PR without necessarily thinking about this implication, we should perhaps remind folk of this once we switch on CD to more repos. I'm not sure if/how much of a problem this might have been for the repos with CD switched on already, if at all, but I think it's worth calling out.
- second, are there any concerns specifically around Dependabot PRs? We have a tendency at present to merge Dependabot PRs without checking a) what's changed in the gem that is being bumped and b) whether this actually works (whether by checking out the branch or even perhaps by looking at integration dashboards). Given that Dependabot PRs have caused some issues in the past (_I don't have links to problem commits but can dig some out_), is there anything else you feel we need to do to ensure Dependabot PRs are good-to-go?

Reply:
> Second, are there any concerns specifically around Dependabot PRs? We have a tendency at present to merge Dependabot PRs without checking a) what's changed in the gem that is being bumped and b) whether this actually works (whether by checking out the branch or even perhaps by looking at integration dashboards). Given that Dependabot PRs have caused some issues in the past (I don't have links to problem commits but can dig some out), is there anything else you feel we need to do to ensure Dependabot PRs are good-to-go?

I think we should try to avoid treating Dependabot PRs as any differently to others. If a Dependabot PR caused an issue with the deployment, then it sounds like we're missing some safety checks in our app/deployment process rather than there being an issue with the PR itself. I suspect that the volume of Dependabot PRs in the past has sometimes led to poor practices (deploying without checking the dashboards as you mention) but I think this RFC would actually help in that respect by mandating the safety checks.

Reply:
Thanks Tom, that sounds reasonable (and I'm definitely not wanting to revisit the previous RFC around Dependabot in my comments here) 👍

Reply:
Thanks for the comments @karlbaker02 and @thomasleese for giving your thoughts. Let's see if we can address them here, or if it would be easier to have a separate discussion:

> it's going to be much easier to press the merge button on a PR without necessarily thinking about this implication

I think you're saying that it's much easier to deploy a breaking change (the button itself should be no more stiff than it is now 😉). This is mentioned later in a consequence: being more cautious. If we're merging stuff late on a Friday now that contains bugs, that blocks any emergency deploys over the weekend. We should stop doing that. I can expand on the consequence and call out this habit as well. Would that address your concern?

> Second, are there any concerns specifically around Dependabot PRs?

I am worried about how I (and others) treat our libraries. @thomasleese is right that there is an issue with how we review Dependabot PRs. Our discussion today focussed on GOV.UK Publishing Components as an example of where we have freely merged Dependabot PRs with breaking changes. I'll recap a few points from that:

- @injms is very keen that we invest in visual regression testing, which makes a lot of sense for this gem. However, the risk for each app is mostly unchanged between manual vs. automatic deployments: if we're not checking the PR manually now, then we will still not be checking the PR manually with CD.

- Although we don't re-test the functionality of the library (that would not be very DRY), the feature tests in our apps do provide some level of reassurance about the functionality. For example, the "button" component must at least render a button, even if it's colour makes it invisible (or something else whacky!)

- Merging a Dependabot PR should really be seen as a failsafe or 2nd chance to halt buggy code in our libraries. In general, we should not be releasing new versions of our libraries without doing some kind of testing with the apps that use them. Once we do a release, it's fair game for someone to assume it will work.

- Our Smoke tests can act as the final "backup" layer for testing that we miss elsewhere. They are far from complete, but they should reassure us that functionality we deem to be critical continues to work.

Despite all that, this RFC does make it more likely that certain kinds of buggy library code will get shipped to production, by reducing the delay between the shipment of the buggy version and its deployment in apps that use it. We must accept that if we want to continue, and I think there's no one way to fix it. However, I think it's worth at least calling it out as a consequence, with some of the above caveats.

How does that sound?

Reply:
Thanks for the explanations Ben, this all sounds good. With the deploying late on a Friday, I was going more along the lines of there being less people around to support if something does goes wrong rather than saying it's going to be easier to deploy a breaking change; deploying a breaking change is probably going to be equivalent between the manual process we have now and any automated CD process.

For the second point about Dependabot PRs, I agree with both yours and Toms comments. One silver lining is that we'll hopefully know sooner if something is an issue through virtue of it getting deployed soon after it's merged, rather than sitting around unreleased, meaning whoever merged should still be around and have context to figure out the best solution.

Reply:
Thanks @karlbaker02! I've done some thinking and have rewritten the consequence around "being more cautious" here: https://github.com/alphagov/govuk-rfcs/pull/128/commits/cecb8f96e6797d6321563f687881459951b4dea5. I'm going to resolve this now, but feel free to comment again if you think there's something still to discuss.